tests/Xml/Encoding/EncodingTest.php

```diff
@@ -263,8 +263,8 @@ public function provideRiskyBidirectionalCases()
             EOXML,
             'data' => ['root' => [
```

Comment:
```php
$xml = \DOM\XMLDocument::createEmpty();

$root = $xml->createElementNS('http://rooty.root', 'root');
$root->setAttributeNS('http://www.w3.org/2000/xmlns/', 'xmlns:test', 'http://testy.test');
$xml->append($root);

$item = $xml->createElement('test:item');
$item->textContent = 'hello';

$root->append($item);

echo $xml->saveXML();exit;

```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root xmlns="http://rooty.root" xmlns:test="http://testy.test">
    <test:item xmlns="">hello</test:item>
</root>
```

The new system is adding an empty `xmlns=""`, even though the `test:item` is the scope.

In the legacy system, this was:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root xmlns="http://rooty.root" xmlns:test="http://testy.test">
    <test:item>hello</test:item>
</root>
```

From an encoding/decoding POV it doesn't make a lot of sense that this information is automatically added.

Not sure how to go forward with this one.

Reply:
The new behaviour is correct. You created the `test:item` element in no namespace, so the serializer outputs `xmlns=""` to make sure that whoever parses the serialized XML knows it should be in no namespace.
I ported the code sample to JavaScript to double check, you can execute this in your browser console and observe the result is the same:

```js
xml = document.implementation.createDocument("", "");
root = xml.createElementNS('http://rooty.root', 'root');
root.setAttributeNS('http://www.w3.org/2000/xmlns/', 'xmlns:test', 'http://testy.test');
xml.append(root);

item = xml.createElement('test:item');
item.textContent = "hello";

root.append(item);

(new XMLSerializer).serializeToString(xml);
```

I suppose you intend to change the namespace of the `test:item` element by adding the xmlns attribute.
In the DOM spec, (even going back to DOM 3 that ext-dom was originally trying to implement), it was never possible to change the namespace of an element. It only appeared to work in "old DOM" because of bugs.

If we take a look at "old DOM" https://3v4l.org/vNAT2 we can see that the prefix and namespace URI are not set because nothing actually changed to the element, it's just that the serialization is wrong.

That being said, I am open to discussing and potentially adding functionality (in a separate class maybe) to manage and manipulate namespace data.

Reply:
> I suppose you intend to change the namespace of the test:item element by adding the xmlns attribute.
In the DOM spec, (even going back to DOM 3 that ext-dom was originally trying to implement), it was never possible to change the namespace of an element. It only gave the impression to work in "old DOM" because of bugs.

This specific example is about encoding a structured array into an XML string.
No manipulations will be done after creating the nodes.
There is also a way to encode into a XMLDocument, on which the user can run additional operations beyond the scope of this encoder.

Example:


```php
        yield 'namespaced' => [
            'xml' => <<<EOXML
                <root xmlns="http://rooty.root" xmlns:test="http://testy.test">
                    <test:item>
                        <id:int xmlns:id="http://identity.id">1</id:int>
                    </test:item>
                </root>
            EOXML,
            'data' => ['root' => [
                '@namespaces' => [
                    '' => 'http://rooty.root',
                    'test' => 'http://testy.test',
                ],
                'test:item' => [
                    'id:int' => [
                        '@namespaces' => [
                            'id' => 'http://identity.id',
                        ],
                        '@value' => '1',
                    ],
                ],
            ]]
        ];
```

In this scenario, the additional `xmlns=""` is rather unexcpected since decoding it would result in a different array than the input array. The idea behind encoding / decoding is that if you perform the same operation, it will result back into the same input (isomorphic). 

This is not a big deal breaker for me, cause it is still valid XML and can be documented from my end.

Reply:
Right, I see the problem. I'm not sure if there's a solution to this, at least I couldn't come up with something immediately.

Reply:
I guess it's fine. I'll just document this behaviour as "it might add additional namespace declarations to make sure the XML is valid".